### ISSUE TITLE
Pass a view to a function

### ISSUE THREAD

**Issue by Squeemos** [skypjack/entt#922]
Hi there,

So the basic idea is I have a scene that holds onto all my entities (and renderables) and I want to pass a view from this into another function that will render them.

They're in different headers, but they all include entt/entt.hpp. The structure looks something like (ignoring everything that's irrelevant)
```
// The struct is included in both headers
struct Example;

// Inside Renderer.h and implemented in Renderer.cpp
class Renderer
{
public:
    void render(const auto& renderables) // This is where I would pass the view
    {
        for (auto e : renderables)
        {
            ....
        }
    }
};

// Inside Scene.h and implemented in Scene.cpp
class Scene
{
public:
    const auto& get_renderables() const
    {
        return registry.view<Example>();
    }
private:
    entt::registry registry;
}
```

And then it's getting called like this:
```
#include "Scene.h"
#include "Renderer.h"
const auto& renderables = scene.get_renderables();
render.render(renderables);
```

But right now the errors I get are:
a function that returns 'auto' cannot be used before it is defined
'renderables' const object must be initialized
'renderables' cannot be used before it is initalized.

I think all the errors stem from it can't determine the type from get_renderables, but I'm not quite sure why.

**Comment by Squeemos** [comment]
My current workaround was passing the registry, but this seems not the best way to go.
```
const ent::registry& get_renderables() const;
```

**Comment by skypjack** [comment]
What's wrong with `entt::view<T, U, PutYourTypesHere>` as return type?
The error is a typical C++ thing, I can't really _fix_ it but you've an easy solution at hand.

**Comment by Squeemos** [comment]
Hmm when I try to do that I get "too many arguments for alias template 'entt::view'" for:
```template <typename T, typename U>
	const entt::view<T, U, Example> get_renderables() const;```
And when I do:
```const entt::view<Transform> get renderables() const;```
I get the error "cannot convert to incomplete class "const entt::basic_view< <error-type>, entt::type_list<>, void>"

I'm sure it is something simple I just can't figure it out haha

**Comment by skypjack** [comment]
It's hard to say without a reproducible example. However, my wild guess is that it's a matter of constness.
From what I see, `get_renderables` is a const method. Therefore, it's likely that it uses a const registry.
In this case, the view can only refer to const types. That is, `entt::view<const Example>`. You can only mix const and non-const types from a non-const registry.

**Comment by Squeemos** [comment]
For sure! I'll include a simple reproducible example below. Perhaps it's how I'm doing my includes or something since when I included everything in a single main.cpp file it compiled and ran just fine.

main.cpp
```
#include <iostream>
#include <string>

#include "Scene.h"
#include "Renderer.h"

int main()
{
	Scene s = Scene();
	Renderer r = Renderer();

	const auto renderables = s.get_renderables();
	r.render(renderables);
	renderables;
	return 0;
} ```
Scene.h
```
#pragma once

#include "entt/entt.hpp"

class Scene
{
public:
	Scene();

	const auto get_renderables() const;

private:
	entt::registry registry;
}; ```
Scene.cpp
```
#include "Scene.h"
#include "Example.h"

Scene::Scene()
{
	entt::entity e = registry.create();
	registry.emplace<Example>(e);
}

const auto Scene::get_renderables() const
{
	return registry.view<Example>();
} ```
Renderer.h
``` #pragma once

#include "entt/entt.hpp"

class Renderer
{
public:
	Renderer() = default;

	void render(const auto renderables);
}; ```
Renderer.cpp
```
#include "Renderer.h"

#include "entt/entt.hpp"

#include <iostream>

void Renderer::render(const auto renderables)
{
	for (auto e : renderables)
	{
		std::cout << static_cast<unsigned int>(e) << std::endl;
	}
} ```
Example.h
```
#pragma once

#include <string>

class Example
{
public:
	Example() : name("Test") {}
private:
	std::string name;
}; ```

**Comment by skypjack** [comment]
To be honest, it really looks like a constness issue. Have you tried it already?
I'm on vacation, so I can't really test your code right away, sorry. I'll do it as soon as I'm back if you didn't solve your issue in the meantime. 👍

**Comment by codylico** [comment]
If I may say something,

Where you have in Scene.h:

```c++
    const auto get_renderables() const;
```

any source file that includes Scene.h can't use that function, since the `auto` type is only resolved in Scene.cpp, which no other source file sees. That leaves two options:
- Make `Scene::get_renderables` inline, so that the function definition within the header resolves the `auto` type, or
- Replace auto with the actual type.

If you went with the second option, you end up with the line you posted earlier:
```c++
const entt::view<Example> get_renderables() const;
```
As for the issue with the incomplete type, you likely need to include "Example.h" within Scene.h, otherwise other source files that include Scene.h don't know what an `Example` is, and therefore don't know what a `entt::view<Example>` is. (Maybe a forward declaration is enough?)

**Comment by Squeemos** [comment]
> To be honest, it really looks like a constness issue. Have you tried it already? I'm on vacation, so I can't really test your code right away, sorry. I'll do it as soon as I'm back if you didn't solve your issue in the meantime. 👍

No worries! I'm in zero rush to get this done thanks for helping while on vacation at all haha.

As for the the suggestions from codylico:
- Tried making it inline, but still defined in the cpp
- Tried making it inline defined in the header
- return entt::view<Example> instead and included Example.h inside Scene.h (since the forward declaration wasn't enough)
- Tried all of the above and removed the const qualifiers

However, I did just fully typed out the view:
```
entt::basic_view<entt::type_list<entt::constness_as_t<entt::storage_type_t<Example, entt::entity>, Example>>, entt::type_list<>, void>
```
and this works! I made a typedef in Scene.h and just included it inside Renderer.h and changed all the function declarations to use this instead. Surely there's gotta be a better way, but I can use this in the meantime.
